### PR TITLE
[Enhance] Show instance statistics before and after through pipeline

### PR DESCRIPTION
**Pull request by JingweiZhang12**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
1. Show all statistics of the dataset: the length of the dataset and the number of instances per category in the dataset.
2. In order to debug in KITTI training, show the variance of the number of instances before and after through the pipeline by modifying the config. Here is an example:
```python
train_dataloader = dict(
    ...,
    dataset=dict(
        type='KittiDataset',
        ...,
        show_ins_var=True))
```

The output information:
```shell
10/17 11:39:38 - mmengine - INFO - ------------------------------                                                                                                                                                                                             
10/17 11:39:38 - mmengine - INFO - The length of the dataset: 3712                                                                                                                                                                                            
10/17 11:39:38 - mmengine - INFO - The number of instances per category in the dataset:
+------------+--------+                                                                                                                                                                                                                                       
| category   | number |            
+------------+--------+                                                                                                                                                                                                                                       
| Pedestrian | 2207   |
| Cyclist    | 734    |                                                                                                        
| Car        | 22158  |
+------------+--------+                                                                                                                                        
10/17 11:43:40 - mmengine - INFO - The number of instances per category after and before through pipeline:
+------------+------------+------------+                                                                                                                                                                                                                      
| category   | new number | ori number |                                                                                                                                                                                                                      
+------------+------------+------------+                                                                                                                                                                                                                      
| Car        | 4          | 4          |                                                                                                                                                                                                                      
| Pedestrian | 12         | 15         |                                                                                                                                                                                                                      
| Cyclist    | 1          | 1          |                                                                                                                                                                                                                      
+------------+------------+------------+
```

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by Tai-Wang** [comment]
Can add an example print in the PR message?

**Review by Tai-Wang — APPROVED** [review]

**Review by ZCMax — APPROVED** [review]

File: mmdet3d/datasets/det3d_dataset.py [hunk 1]
```diff
@@ -58,6 +59,8 @@ class Det3DDataset(BaseDataset):
             which can be used in Evaluator. Defaults to True.
         file_client_args (dict, optional): Configuration of file client.
             Defaults to dict(backend='disk').
+        show_ins_statistics (bool, optional): Whether to show the statistics of
```
Comment (ZwwWayne) [inline mmdet3d/datasets/det3d_dataset.py:62]: We can set its default to True
show_ins_statistics -> show_statistics

File: mmdet3d/datasets/det3d_dataset.py [hunk 2]
```diff
@@ -318,12 +326,34 @@ def prepare_data(self, index: int) -> Optional[dict]:
...
         example = self.pipeline(input_dict)
+
         if not self.test_mode and self.filter_empty_gt:
             # after pipeline drop the example with empty annotations
             # return None to random another in `__getitem__`
             if example is None or len(
                     example['data_samples'].gt_instances_3d.labels_3d) == 0:
                 return None
+
+        if self.show_ins_statistics:
```
Comment (ZwwWayne) [inline mmdet3d/datasets/det3d_dataset.py:337]: 1. Wrap it into a callable function.
2. We can simply do it once after the data is loaded. 
3. Show the information dataset length, class names, and number of instances of each class

File: mmdet3d/datasets/det3d_dataset.py [hunk 3]
```diff
@@ -318,12 +326,34 @@ def prepare_data(self, index: int) -> Optional[dict]:
...
+            new_num_per_cat = dict()
+            for label in example['data_samples'].gt_instances_3d.labels_3d:
+                cat_name = self.metainfo['CLASSES'][label]
+                new_num_per_cat[cat_name] = new_num_per_cat.get(cat_name,
+                                                                0) + 1
+            content_show = ''
+            for cat_name, num in ori_num_per_cat.items():
+                new_num = new_num_per_cat.get(cat_name, 0)
+                content_show += f'\n{cat_name}'.ljust(25) + f'{new_num}/{num}'
+            logger.info(
```
Comment (VVsssssk) [inline mmdet3d/datasets/det3d_dataset.py:353]: I think you can add some simplified prefixes for print info to distinguish dataset info and data info. And maybe you need to modify [dbsampler info](https://github.com/open-mmlab/mmdetection3d/blob/c9541b0db89498fdea5cafd05b7b17f7b625b858/mmdet3d/datasets/pipelines/dbsampler.py#L139) to add a prefix with different with other info to make it more clear.

File: mmdet3d/datasets/det3d_dataset.py (hunk 1 above)
Comment (JingweiZhang12) [reply mmdet3d/datasets/det3d_dataset.py:62]: Considering we show all statistics of the dataset by default, maybe show_ins_num_var is better if we only want it to be a debugging switch.

**Review by VVsssssk — APPROVED** [review]

**Comment by ZwwWayne** [comment]
Suggest rebasing the code to the most recent dev.

File: mmdet3d/datasets/det3d_dataset.py [hunk 4]
```diff
@@ -58,6 +60,9 @@ class Det3DDataset(BaseDataset):
             which can be used in Evaluator. Defaults to True.
         file_client_args (dict, optional): Configuration of file client.
             Defaults to dict(backend='disk').
+        show_ins_num_var (bool, optional): Whether to show variance of the
```
Comment (ZwwWayne) [inline mmdet3d/datasets/det3d_dataset.py:63]: show_ins_variance:  For debug purpose. Whether to show the variance of the number of instances....

File: mmdet3d/datasets/det3d_dataset.py [hunk 5]
```diff
@@ -125,6 +138,21 @@ def __init__(self,
         self.metainfo['box_type_3d'] = box_type_3d
         self.metainfo['label_mapping'] = self.label_mapping
 
+        # used for showing variance of the number of instances before and
+        # after through the pipeline
+        self.show_ins_num_var = show_ins_num_var
+
+        # show statistics of this dataset
```
Comment (ZwwWayne) [inline mmdet3d/datasets/det3d_dataset.py:145]: We can simply build a table for better visualization

File: mmdet3d/datasets/det3d_dataset.py [hunk 6]
```diff
@@ -105,13 +112,20 @@ def __init__(self,
             for label_idx, name in enumerate(metainfo['CLASSES']):
                 ori_label = self.METAINFO['CLASSES'].index(name)
                 self.label_mapping[ori_label] = label_idx
+
+            self.ins_num_per_cat = {name: 0 for name in metainfo['CLASSES']}
```
Comment (ZwwWayne) [inline mmdet3d/datasets/det3d_dataset.py:116]: ins_num_per_cat ->num_ins_per_cat

File: mmdet3d/datasets/det3d_dataset.py [hunk 7]
```diff
@@ -58,6 +61,9 @@ class Det3DDataset(BaseDataset):
             which can be used in Evaluator. Defaults to True.
         file_client_args (dict, optional): Configuration of file client.
             Defaults to dict(backend='disk').
+        show_ins_var (bool, optional): For debug purpose. Whether to show
+            variance of the number of instances before and after through
```
Comment (ZwwWayne) [inline mmdet3d/datasets/det3d_dataset.py:65]: variance--> variation

File: mmdet3d/datasets/det3d_dataset.py [hunk 8]
```diff
@@ -125,7 +139,23 @@ def __init__(self,
         self.metainfo['box_type_3d'] = box_type_3d
         self.metainfo['label_mapping'] = self.label_mapping
 
-    def _remove_dontcare(self, ann_info: dict) -> dict:
+        # used for showing variance of the number of instances before and
+        # after through the pipeline
+        self.show_ins_var = show_ins_var
+
+        # show statistics of this dataset
+        logger: MMLogger = MMLogger.get_current_instance()
```
Comment (ZwwWayne) [inline mmdet3d/datasets/det3d_dataset.py:147]: better to use print_log(xxx, 'current')

File: mmdet3d/datasets/det3d_dataset.py [hunk 9]
```diff
@@ -291,6 +326,32 @@ def parse_data_info(self, info: dict) -> dict:
 
         return info
 
+    def _show_ins_num_var(self, old_labels: np.ndarray,
```
Comment (ZwwWayne) [inline mmdet3d/datasets/det3d_dataset.py:329]: _shoe_ins_var

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**